File: src/camera_tools/src/take_picture_server.cpp
//
// Create by Tim Mascal on 1/9/25.
//

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>
#include "camera_tools_interfaces/srv/take_picture.hpp" // Update to use your service definition

using namespace std::placeholders;

class TakePictureServiceServer : public rclcpp::Node
{
public:
    TakePictureServiceServer()
        : Node("take_picture_service_server")
    {
        service_ = this->create_service<camera_tools_interfaces::srv::TakePicture>(
            "take_picture",
            std::bind(&TakePictureServiceServer::handle_request, this, _1, _2));
        RCLCPP_INFO(this->get_logger(), "Service server ready.");

        // Video capturing logic
        cap.open(2, cv::CAP_V4L2); // Open the default camera (index 0)

        if (!cap.isOpened()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open the camera.");
            return;
        }

        // Set the desired resolution
        int desired_width = 2592;  // Hardcoded width
        int desired_height = 1944; // Hardcoded height
        cap.set(cv::CAP_PROP_FRAME_WIDTH, desired_width);
        cap.set(cv::CAP_PROP_FRAME_HEIGHT, desired_height);
        // Attempt to enable camera auto-focus
        if (cap.set(cv::CAP_PROP_AUTOFOCUS, 1)) {
            RCLCPP_INFO(this->get_logger(), "Auto-focus enabled.");
        } else {
            RCLCPP_WARN(this->get_logger(), "Auto-focus not supported by the camera.");
        }
    }

private:
    void handle_request(const std::shared_ptr<camera_tools_interfaces::srv::TakePicture::Request> request,
                    std::shared_ptr<camera_tools_interfaces::srv::TakePicture::Response> response)
{
    (void)request;
    // Discard initial frames to ensure the captured frame is focused
    for (int i = 0; i < 5; ++i) {
        cap >> frame;
    }

    // Delay to allow the camera's auto-focus to adjust
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    cap >> frame; // Capture and discard frame

    if (frame.empty()) {
        RCLCPP_ERROR(this->get_logger(), "Captured frame is empty.");
        return;
    }

    // Convert cv::Mat to ROS2 sensor_msgs/Image
    auto img_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", frame).toImageMsg();
    response->image = *img_msg;

    RCLCPP_INFO(this->get_logger(), "Image captured and returned.");
}

    // Declare a service Server for TakePicture Service
    rclcpp::Service<camera_tools_interfaces::srv::TakePicture>::SharedPtr service_;

    // Declare member variables for video capturing and frame storage
    cv::VideoCapture cap;
    cv::Mat frame;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TakePictureServiceServer>());
    rclcpp::shutdown();
    return 0;
}

File: src/camera_tools/scripts/take_picture_server.py
#!/home/mark/ros2-humble-env/bin/python3
# Created by Tim Mascal on 2/23/25

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
from camera_tools_interfaces.srv import TakePicture  # Update with your service definition


class TakePictureServiceServer(Node):
    def __init__(self):
        super().__init__('take_picture_service_server')
        # Create the service
        self.service_ = self.create_service(
            TakePicture,
            'take_picture',
            self.handle_request
        )
        # Initialize CvBridge for image conversion
        self.bridge = CvBridge()
        self.get_logger().info('Service server ready.')

    def handle_request(self, request, response):
        """Handle service requests"""
        # Open the camera
        cap = cv2.VideoCapture(2, cv2.CAP_V4L2)

        if not cap.isOpened():
            self.get_logger().error('Failed to open the camera.')
            return response

        # Set the desired resolution
        desired_width = 2592  # Hardcoded width
        desired_height = 1944  # Hardcoded height
        cap.set(cv2.CAP_PROP_FRAME_WIDTH, desired_width)
        cap.set(cv2.CAP_PROP_FRAME_HEIGHT, desired_height)

        # Capture a frame
        ret, frame = cap.read()
        cap.release()  # Release the camera resource after capturing

        if not ret or frame is None:
            self.get_logger().error('Captured frame is empty.')
            return response

        # Convert the captured frame (OpenCV image) to ROS2 sensor_msgs/Image
        img_msg = self.bridge.cv2_to_imgmsg(frame, encoding="bgr8")
        response.image = img_msg

        self.get_logger().info('Image captured and returned.')
        return response


def main(args=None):
    rclpy.init(args=args)

    service_server = TakePictureServiceServer()
    try:
        rclpy.spin(service_server)
    except KeyboardInterrupt:
        pass
    finally:
        service_server.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()